spellmonitor.lic

```diff
@@ -43,139 +40,121 @@ class DRSpells
+#
+# One or more spells may be listed between a <pushStream/> <popStream/> pair,
+# but only one spell and its duration are ever listed per line.
+#
+spell_action = proc do |server_string|
+  # New information about active spells, reset our cached data.
+  if server_string =~ %r{<clearStream id="percWindow"/>}
+    DRSpells.slivers = false
+    DRSpells.refresh_data.each_key { |k| DRSpells.refresh_data[k] = false }
+    begin
+      Thread.new do
+        sleep 0.5
+        DRSpells.refresh_data.each do |key, state|
+          unless state
+            DRSpells.active_spells.delete(key)
+            DRSpells.refresh_data.delete(key)
+          end
+        end
+      end
+    rescue StandardError => e
+      echo(e)
+      echo('Error in spell monitor')
+    end
+    next server_string
+  end
+
+  # Abilities may be prefixed or suffixed by
+  # a control character that appears like whitespace
+  # but isn't whitespace so 'strip' doesn't remove it.
+  # Replace all instances of codepoint 32 with ' '.
+  data = server_string.dup.gsub(32.chr, ' ').strip
+
+  # Indicates we're done with the prior <pushStream/> tag.
+  # However, the next characters in the line might be
+  # another <pushStream/> or something else, so keep going.
+  if data =~ %r{^<popStream/>}
+    DRSpells.grabbing_spell_data = false
+    data.slice!(0, 12).strip!
+  end
+
+  # The start of one or more active spells!
+  if data =~ %r{^<pushStream id="percWindow"/>}
+    DRSpells.grabbing_spell_data = true
+    data.slice!(0, 29).strip!
+  end
+
+  if DRSpells.grabbing_spell_data
+    spell = nil
+    duration = nil
+    case data
+    when /(?<spell>[^<>]+?)\s+\((?:\D*)(?<duration>\d+)\s*(?:%|roisae?n)\)/
+      # Spell with known duration remaining
+      spell = Regexp.last_match[:spell]
+      duration = Regexp.last_match[:duration].to_i
+    when /(?<spell>[^<>]+?)\s+\(Fading\)/
+      # Spell fading away
+      spell = Regexp.last_match[:spell]
+      duration = 0
+    when /(?<spell>[^<>]+?)\s+\(Indefinite|OM\)/
+      # Cyclic spell or Osrel Meraud cyclic spell
+      spell = Regexp.last_match[:spell]
+      duration = UNKNOWN_DURATION
+    when /(?<spell>[^<>]+?)\s+\(.+\)/
+      # Spells with inexact duration verbiage, such as with
+      # Barbarians without knowledge of Power Monger mastery
+      spell = Regexp.last_match[:spell]
+      duration = UNKNOWN_DURATION
+    when /.*orbiting sliver.*/
+      # Moon Mage slivers
+      DRSpells.slivers = true
+    when /^(.*)$/
+      # No idea what we received, just a general catch all
+      spell = Regexp.last_match(1)
+      duration = UNKNOWN_DURATION
+    end
+    if spell
+      DRSpells.active_spells[spell] = duration
+      DRSpells.refresh_data[spell] = true
+    end
+  end
+  server_string
+end
+
+# Unsure the historical significance of this pause
+# https://github.com/rpherbig/dr-scripts/commit/04b7474b21c809d960ccb0c720808b44ae0f1ff8
+pause 5
```

Comment:
I went through the commit history but couldn't determine the specific reason for pausing 5 seconds before wiring up the downstream hooks

Reply:
This is done due to a race condition sometimes not hitting before this triggers.  You will get an issue where spells never update because they use different xml. Your fix, if consolidating the parsing into a single will eliminate the need for this.

Reply:
Gotcha, thanks, Jay